src/users/user.service.ts

```diff
@@ -40,6 +41,59 @@ export class UserService {
     return { data: users, total };
   }
 
+  async getUserBySearch(searchUserDto: SearchUserDto): Promise<{ data: User[]; total: number }> {
+    let users = [];
+    let optionalQuery = {};
+    let splitName: string[];
+    let qFirstName, qLastName, qName, qEmail: RegExp;
+    const { user_id, email, name, limit = 0, offset = 0, sort, desc } = searchUserDto;
+    const sorting = sort ? { [sort]: desc } : { createdAt: -1 };
+    // eslint-disable-next-line prefer-const
+    qEmail = new RegExp(`.*${email}.*`, "i");
+    if (name && name.includes(" ")) {
+      splitName = name.split(" ");
+      qFirstName = new RegExp(`.*${splitName[0]}.*`, "i");
+      qLastName = new RegExp(`.*${splitName[1]}.*`, "i");
+      optionalQuery = {
+        $and: [{ firstName: qFirstName }, { lastName: qLastName }],
+      };
+    }
+    if (name && !name.includes(" ")) {
+      qName = new RegExp(`.*${name}.*`, "i");
+      if (!email && !user_id) {
+        optionalQuery = {
+          $or: [{ firstName: qName }, { lastName: qName }],
+        };
+      } else {
+        user_id.match(/^[0-9a-fA-F]{24}$/)
+          ? (optionalQuery = {
+              $or: [{ firstName: qName }, { lastName: qName }, { _id: user_id }, { email: qEmail }],
+            })
+          : (optionalQuery = {
+              $or: [{ firstName: qName }, { lastName: qName }, { email: qEmail }],
+            });
+      }
+    }
+    if (user_id && !email && !name) {
+      user_id.match(/^[0-9a-fA-F]{24}$/)
+        ? (optionalQuery = { _id: user_id })
+        : (optionalQuery = { email: user_id + "@wrong" });
```

Comment:
what does this mean?

Reply:
Different from "FindOne" or "FindById" method in MongoDb, if I do not "Find" the _id field by the string of a standard 12-byte ObjectId, there would be a 500 error. So I just search a safe text that would return empty result before. Now I would just return out of the function when user_id is not 12-byte string.